<launch>
  
  <arg name="input" default="/pcl_stream" />

  <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen" />

  <!-- Dropping -->
  <!-- Kinect runs around 7hz - dropping 6 out of 7 gets us to ~1hz -->
  <arg name="drop_x" default="6" />
  <arg name="drop_y" default="7" />
  <node name="pipeline_throttle" pkg="pcl_pipeline_utils" type="dropper.py" output="screen"
    args="$(arg input) $(arg drop_x) $(arg drop_y) /pipeline/01_throttled" />

  <!-- Decimation -->
  <arg name="decimation_leafsize_m" default="0.01" />
  <node pkg="nodelet" type="nodelet" name="pipeline_decimation" args="load pcl/VoxelGrid pcl_manager" output="screen">
    <remap from="~input" to="$(arg input)" />

    <param name="filter_field_name" type="string" value="z" />
    <param name="filter_limit_min" type="double" value="0.4" />
    <param name="filter_limit_max" type="double" value="5" />
    <param name="filter_limit_negative" type="bool" value="false" />
    <param name="leaf_size" type="double" value="$(arg decimation_leafsize_m)" />

    <remap from="~output" to="/pipeline/02_decimated" />
  </node>

  <!-- Outlier Removal -->
  <arg name="outlier_mean_k" default="10" />
  <arg name="outlier_stddev" default="1.0" />
  <node pkg="nodelet" type="nodelet" name="pipeline_outlier_removal" output="screen"
    args="load pcl/StatisticalOutlierRemoval pcl_manager" >
    <remap from="~input" to="/pipeline/02_decimated" />

    <param name="mean_k" type="int" value="$(arg outlier_mean_k)" />
    <param name="stddev" type="double" value="$(arg outlier_stddev)" />
    <param name="negative" type="bool" value="false" />

    <remap from="~output" to="/pipeline/03_outliers_removed" />
  </node>

  <!-- Compression -->
  <arg name="compression_profile" default="LOW_RES_ONLINE_COMPRESSION_WITH_COLOR" />
  <node pkg="pcl_pipeline_utils" type="pcl_compression" name="pipeline_compression" output="screen" >
    <remap from="input" to="/pipeline/03_outliers_removed" />

    <param name="profile" type="string" value="$(arg compression_profile)" />

    <remap from="output" to="/pipeline/04_compressed" />
  </node>

  <!-- Decompression -->
  <node pkg="pcl_pipeline_utils" type="pcl_decompression" name="pipeline_decompression" output="screen" >
    <remap from="input" to="/pipeline/04_compressed" />
    <remap from="output" to="/pipeline/05_decompressed" />
  </node>
  
  <!-- MLS Smoothing -->
  <arg name="mls_search_radius" default="0.3" />
  <node pkg="pcl_pipeline_utils" type="pcl_smoothing" name="pipeline_smoothing" output="screen" >
    <remap from="input" to="/pipeline/05_decompressed" />

    <param name="search_radius" type="double" value="$(arg mls_search_radius)" />

    <remap from="output" to="/pipeline/06_smoothed" />
  </node>

  <!-- Meshing -->
  <arg name="meshing_method" default="GreedyProjection" />
  <arg name="meshing_greedy_search_radius" default="2.5" />
  <arg name="meshing_greedy_mu" default="2.5" />
  <arg name="meshing_greedy_max_nearerst_neighbours" default="500" />
  <arg name="meshing_greedy_max_surface_angle" default="45" />
  <arg name="meshing_greedy_min_surface_angle" default="10" />
  <arg name="meshing_greedy_max_angle" default="120" />
  <arg name="meshing_poisson_depth" default="9" />
  <node pkg="pcl_pipeline_utils" type="pcl_meshing" name="pipeline_meshing" output="screen" >
    <remap from="input" to="/pipeline/05_decompressed" />

    <param name="method" type="string" value="$(arg meshing_method)" />
    <param name="greedy_search_radius" type="double" value="$(arg meshing_greedy_search_radius)" />
    <param name="greedy_search_radius" type="double" value="$(arg meshing_greedy_mu)" />
    <param name="greedy_search_radius" type="int" value="$(arg meshing_greedy_max_nearerst_neighbours)" />
    <param name="greedy_search_radius" type="double" value="$(arg meshing_greedy_max_surface_angle)" />
    <param name="greedy_search_radius" type="double" value="$(arg meshing_greedy_min_surface_angle)" />
    <param name="greedy_search_radius" type="double" value="$(arg meshing_greedy_max_angle)" />
    <param name="poisson_depth" type="int" value="$(arg meshing_poisson_depth)" />

    <remap from="output" to="/pipeline/07_meshed" />
  </node>

  <!-- UnMeshing -->
  <node pkg="pcl_pipeline_utils" type="pcl_unmeshing" name="pipeline_unmeshing" output="screen" >
    <remap from="input" to="/pipeline/07_meshed" />
    <remap from="output" to="/pipeline/08_unmeshed" />
  </node>
  

</launch>